### PR TITLE
Patch from CKAN master Fix for Harvest Source Form

### LINKED ISSUE #75: [BUG] Unable to create waf-collection harvest source
## Notice
This issue is blocking [Sandblox-PR#1792](https://github.com/GSA/datagov-deploy/pull/1792). 

## Error
We can't create `waf-collection` harvest source in the _new harvest source_ form

image

This is because we are not using the custom form to create this type of source.
The catalog classic use `datagov_harvest` instead `harvest` in the plugin list. This is defined at [geodatagov extension](https://github.com/GSA/ckanext-geodatagov/blob/a674cb96a6cacacfcc67d992845159dff55682d4/setup.py#L30). This class [override](https://github.com/GSA/ckanext-geodatagov/blob/eee214455065e775628508cc6d06d5d16c1893b2/ckanext/geodatagov/plugins.py#L309) the form to create harvest sources.

If we move plugin from `harvest` to `datagov_harvest` we still have an error

image

## How to reproduce

1. Replace the `harvest` extension with `datagov_harvest` in the plugins configuration (production.ini)
1. Open the new harvest form https://catalog-next.sandbox.datagov.us/harvest/new
1. Fill in the details for a waf-collection harvester
1. Click Save

## Tasks

- [ ] Define if want to use `datagov_harvest` and fix the error or move this new form (and the other features) to a different extension preserving the `harvest` plugin in use. ([discussion here](https://github.com/GSA/datagov-deploy/pull/1792#issuecomment-648536304))
 - [ ] Write issues regarding this problem

## Notes

This new plugin [was added](https://github.com/GSA/ckanext-geodatagov/commit/132f5a8f04340bdb7a91dd15f1a3f1913d2489d6) for Z3950 harvester 7 years ago.

The order of plugins is different:
Screen Shot 2020-06-24 at 10 56 51 AM

## Logs

Using `datagov_harvest` we see this error log:
 
```
Module ckan.logic.action.create:170 in package_create         

 
check_data_dict	None
context	{'__auth_audit': [], '__auth_user_obj_checked': True, 'auth_user_obj': <User id=6fb43a58-a883-47c b-8e19-8503d7f32064 name=admin password=$pbkdf2-sha512$25000$WgvhHOOc897bO6dUau09xw$waMJ6mmbgBynQpo8aep6ijGRm8EqijCjlQ5aeMPqA5XfjxDKE.QSXO8czsiBHUnEisH8oxC8QutjPWgItmFsig fullname=None email=your_email@example.com apikey=f1bb38d5-8841-48cb-a414-cda128f811ab created=2020-06-23 15:39:44.994457 reset_key=None about=None activity_streams_email_notifications=False sysadmin=True state=active>, 'message': '', 'model': <module 'ckan.model' from '/srv/app/src/ckan/ckan/model/__init__.pyc'>, 'save': True, 'session': <sqlalchemy.orm.scoping.scoped_session object at 0x7f292908ef50>, 'user': u'admin'}

data_dict	{'collection_metadata_url': u'https://meta.geo.census.gov/data/existing/dec ennial/GEO/GPMB/TIGERline/TIGER2018/SeriesInfo/SeriesCollection_tl_2018_concity.shp.iso.xml', 'database': u'', 'extra_search_criteria': u'', 'extras': [], 'frequency': u'MANUAL', 'name': u'gpb-5', 'notes': u'', 'owner_org': u'1c4b46eb-afe8-4d4c-a974-822f9d7df76a', 'port': u'', 'private_datasets': u'False', 'save': u'Save', 'source_type': u'waf-collection', 'title': u'GPB 5b', 'type': 'harvest', 'url': u'https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/concity/', 'validator_profiles': u'', 'validator_schema': u''}

model	<module 'ckan.model' from '/srv/app/src/ckan/ckan/model/__init__.pyc'>
package_plugin	<Plugin DataGovHarvest 'datagov_harvest'>

schema	{'__extras': [<function harvest_source_extra_validator at 0x7f29261ceed8>], 'config': [<fun ction ignore_missing at 0x7f29263740c8>, <function harvest_source_config_validator at 0x7f29261cede8>, <function convert_to_extras at 0x7f29262bcde8>], 'extras': {'__extras': [<function ignore at 0x7f292636cf50>], 'deleted': [<function ignore_missing at 0x7f29263740c8>], 'id': [<function ignore at 0x7f292636cf50>], 'key': [<function not_empty at 0x7f292636cd70>, <function extra_key_not_in_root_schema at 0x7f29274fe578>, <function unicode_safe at 0x7f29263742a8>], 'revision_timestamp': [<function ignore at 0x7f292636cf50>], 'state': [<function ignore at 0x7f292636cf50>], 'value': [<function not_missing at 0x7f292636ccf8>]}, 'frequency': [<function ignore_missing at 0x7f29263740c8>, <type 'unicode'>, <function harvest_source_frequency_exists at 0x7f29261d60c8>, <function convert_to_extras at 0x7f29262bcde8>], 'name': [<function not_empty at 0x7f292636cd70>, <type 'unicode'>, <function name_validator at 0x7f29274fd500>, <function package_name_validator at 0x7f29274fd578>], 'notes': [<function ignore_missing at 0x7f29263740c8>, <type 'unicode'>], 'organization': [<function ignore_missing at 0x7f29263740c8>], 'owner_org': [<function owner_org_validator at 0x7f29274eeb90>, <type 'unicode'>], 'private': [<function ignore_missing at 0x7f29263740c8>, <function boolean_validator at 0x7f29274eede8>, <function datasets_with_no_organization_cannot_be_private at 0x7f29274fe230>], 'save': [<function ignore at 0x7f292636cf50>], 'source_type': [<function not_empty at 0x7f292636cd70>, <type 'unicode'>, <function harvest_source_type_exists at 0x7f29261ced70>, <function convert_to_extras at 0x7f29262bcde8>], 'state': [<function ignore_missing at 0x7f29263740c8>], 'title': [<function callable at 0x7f292127e7d0>, <type 'unicode'>], 'type': [<function dataset_type_exists at 0x7f29261d6140>, <type 'unicode'>], 'url': [<function not_empty at 0x7f292636cd70>, <type 'unicode'>, <function harvest_source_url_validator at 0x7f29261cecf8>]}

user	u'admin'



Module ckan.lib.navl.dictization_functions:452 in unflatten         

 
convert_to_list	[]
current_pos	[{'key': 'frequency', 'value': u'MANUAL'}]
data	{
    ('__extras',): {
        'collection_metadata_url': u'https://meta.geo.census.gov/data/e xisting/decennial/GEO/GPMB/TIGERline/TIGER2018/SeriesInfo/SeriesCollection_tl_2018_concity.shp.iso.xml', 
        'database': u'', 
        'extra_search_criteria': u'', 
        'extras': [], 
        'port': u'', 
        'private_datasets': u'False', 
        'validator_profiles': u'', 
        'validator_schema': u''
}, 
    ('collection_metadata_url',): u'https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/SeriesInfo/SeriesCollection_tl_2018_concity.shp.iso.xml', 
    ('config',): '{
        "collection_metadata_url": "https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/SeriesInfo/SeriesCollection_tl_2018_concity.shp.iso.xml", 
        "private_datasets": false
    }', 
    ('extras',): [
        {'key': 'frequency', 'value': u'MANUAL'}
        ], 
    ('extras', 0, 'key'): 'frequency', ('extras', 0, 'value'): u'MANUAL', 
    ('extras', 1, 'key'): 'source_type', ('extras', 1, 'value'): u'waf-collection', 
    ('frequency',): u'MANUAL', 
    ('name',): u'gpb-5', 
    ('notes',): u'', 
    ('owner_org',): u'1c4b46eb-afe8-4d4c-a974-822f9d7df76a', 
    ('private_datasets',): False, 
    ('source_type',): u'waf-collection', 
    ('title',): u'GPB 5b', 
    ('type',): u'harvest', 
    ('url',): u'https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/concity/'
}

flattend_key	('extras', 1, 'key')
key	1

unflattened	{'__extras': {'collection_metadata_url': u'https://meta.geo.census.gov/data/exis ting/decennial/GEO/GPMB/TIGERline/TIGER2018/SeriesInfo/SeriesCollection_tl_2018_concity.shp.iso.xml', 'database': u'', 'extra_search_criteria': u'', 'extras': [], 'port': u'', 'private_datasets': u'False', 'validator_profiles': u'', 'validator_schema': u''}, 'collection_metadata_url': u'https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/SeriesInfo/SeriesCollection_tl_2018_concity.shp.iso.xml', 'config': '{"collection_metadata_url": "https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/SeriesInfo/SeriesCollection_tl_2018_concity.shp.iso.xml", "private_datasets": false}', 'extras': [{'key': 'frequency', 'value': u'MANUAL'}], 'frequency': u'MANUAL', 'name': u'gpb-5', 'notes': u'', 'owner_org': u'1c4b46eb-afe8-4d4c-a974-822f9d7df76a', 'private_datasets': False, 'source_type': u'waf-collection', 'title': u'GPB 5b', 'type': u'harvest', 'url': u'https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/concity/'}
view
>>  current_pos = current_pos[key]
IndexError: list index out of range
```

### PR DESCRIPTION
**Pull request by avdata99**
Related to issue #75.

 - CKAN at master have a patch for this. Waiting for the patch to be backported to CKAN 2.8
 - Moving to `datagov_harvest` plugin fixes #75 but harvest sources still fails to edit after they are created.

### Patch

[10 days ago](https://github.com/ckan/ckan/commit/1e573fce60e2ae7169fa9811c2e080ab7e02a883) this function was improved. This PR maybe fix this error. This is part of [a PR](https://github.com/ckan/ckan/pull/5444) which is marked to _backport_ in next days.
**Propose: add a patch and update to CKAN 2.8.5 when becoming available.**

### Manual test

We create a `waf-collection` source:
 - URL: https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/concity/
 - type: WAF homegeneous collection
 - validator: ISO 19115
 - Collection Metadata Url: https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/SeriesInfo/SeriesCollection_tl_2018_concity.shp.iso.xml

Saving harvest source **now is working** (seems fixed) **but if I try to edit** this source the `Collection Metadata Url` field is missing.

### CODE REVIEW

File: ckan/patches/ckan/unflattern_indexerror.patch [hunk 1]
```diff
@@ -0,0 +1,39 @@
+diff --git a/ckan/lib/navl/dictization_functions.py b/ckan/lib/navl/dictization_functions.py
```
Comment (adborden) [inline ckan/patches/ckan/unflattern_indexerror.patch:1]: Can you use `git format-patch` to create this diff instead? That way the commit and author information can be preserved as part of the patch.

File: ckan/patches/ckan/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+### Test
+
+We create a `waf-collection` source:
+ - URL: https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/concity/
+ - type: WAF homegeneous collection
+ - validator: ISO 19115
+ - Collection Metadata Url: https://meta.geo.census.gov/data/existing/decennial/GEO/GPMB/TIGERline/TIGER2018/SeriesInfo/SeriesCollection_tl_2018_concity.shp.iso.xml
+
+Save it's workings (seems fixed) but if I try to edit this source the `Collection Metadata Url` field is missing.
```
Comment (adborden) [inline ckan/patches/ckan/README.md:26]: This is great, can we write an automated test around this?

Comment (avdata99) [reply ckan/patches/ckan/README.md:26]: > This is great, can we write an automated test around this?

We added a test [here](https://github.com/GSA/ckanext-geodatagov/blob/75_fix_harvest_source_edit/ckanext/geodatagov/tests/test_source_form.py#L21)

**Comment by avdata99** [comment]
> I think this looks good - we need to look into why the database fields are showing up here and not in production. It would be nice to have a test for this as well

I added a test for `database` and `port` fields

File: ckan/patches/ckan/unflattern_indexerror.patch (hunk 1 above)
Comment (avdata99) [reply ckan/patches/ckan/unflattern_indexerror.patch:10]: Updated [here](https://github.com/GSA/catalog.data.gov/pull/77/commits/58846679cd8f833a744d90c859e6829392489732)

File: ckan/patches/ckan/README.md (hunk 2 above)
Comment (adborden) [reply ckan/patches/ckan/README.md:26]: I'd like to see a BATS test, if possible. That way the test for the patch lives in the same repo as the patch itself.

Looking at your description of the test above, it looks like we just need to POST to the /dataset/new form with the waf-collection parameters. Sounds like a straightforward test. Is it possible?

Comment (avdata99) [reply ckan/patches/ckan/README.md:26]: [Here](https://github.com/GSA/catalog.data.gov/pull/77/commits/a99647d470af4f75603b5be05327bbd2c51d8786) we have a test for this

File: tests/test-base.bats [hunk 3]
```diff
@@ -59,6 +59,27 @@ load test_helper
...
 
+@test "Harvest create form" {
+  # create a waf-collection harvest source
+  # require the organization created in previous test
+
+  local api_key json_data existing_org
+  api_key=$(db -c "select apikey from public.user where name='$CKAN_SYSADMIN_NAME';")
+  json_data=$(sed s/\$RNDCODE/$RNDCODE/g /tests/test-data/waf-collection-source.json)
+
+  run curl --silent -X POST \
```
Comment (adborden) [inline tests/test-base.bats:70]: Probably worth a helper for API requests.

File: tests/test-base.bats [hunk 4]
```diff
@@ -59,6 +59,27 @@ load test_helper
   assert_json .success true
 }
 
+@test "Harvest create form" {
+  # create a waf-collection harvest source
+  # require the organization created in previous test
```
Comment (adborden) [inline tests/test-base.bats:64]: Can we create (and destroy) the organization as part of the test setup (and cleanup)? I'm not a fan of these "stateful" tests where one test depends on the previous one.

File: tests/test-base.bats [hunk 5]
```diff
@@ -59,6 +59,27 @@ load test_helper
   assert_json .success true
 }
 
+@test "Harvest create form" {
+  # create a waf-collection harvest source
```
Comment (adborden) [inline tests/test-base.bats:63]:
```suggestion
  # create a waf-collection harvest source
  # asserts ckan/patches/ckan/unflattern_indexerror.patch is applied
```

File: tests/test-data/waf-collection-source.json [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{
+    "name": "waf-collection-source-$RNDCODE",
+    "url": "http://test.com-$RNDCODE",
+    "source_type": "waf-collection",
+    "title": "WAF test $RNDCODE",
+    "collection_metadata_url": "http://coll.test.com-$RNDCODE",
+    "frequency": "MANUAL",
+    "owner_org": "test-organization-$RNDCODE"
+}
```
Comment (adborden) [inline tests/test-data/waf-collection-source.json:9]: With such a few lines, I think it's easier to read if this was just inline within the test.

File: tests/test-base.bats [hunk 7]
```diff
@@ -59,6 +59,27 @@ load test_helper
   assert_json .success true
 }
 
+@test "Harvest create form" {
```
Comment (adborden) [inline tests/test-base.bats:62]: More descriptive test descriptions, please.

```suggestion
@test "Given an organization, a waf-collection harvest source is created successfully from form" {
```

**Comment by avdata99** [comment]
> Nice simple test, thanks!
> 
> I am taking this opportunity to pick on some testing practices, more for discussion rather than requesting fixes :)
> 
> One major question though, I'm surprised to see the `package_create` API under test instead of the form endpoint. I thought #75 was specific to the form, and not the package_create action. Can you clarify?

I'm also surprised. It's a mistake :(

File: tests/test_helper.bash [hunk 8]
```diff
@@ -152,6 +152,95 @@ function test_extension_loaded() {
...
+    -H 'cache-control: no-cache' \
+    -H 'content-type: application/json' \
+    -d "$json_data"
+
+  [ "$status" = 0 ]
+  
+  # only expect JSON results for POST calls
+  assert_json .success true
+  
+  api_results=$output
```
Comment (adborden) [inline tests/test_helper.bash:199]: Why the extra variable? Why not just use stdout or $output in the caller?

**Pull request merged**